framework/src/media/Decoder.cpp

```diff
@@ -29,42 +30,46 @@ Decoder::Decoder(const Decoder* source)
 Decoder::~Decoder()
 {
 #ifdef CONFIG_AUDIO_CODEC
-	if (pv_player_finish(&mPlayer) != PV_SUCCESS) {
-		meddbg("Error! pv_player_finish failed!\n");
+	if (mPlayer) {
+		if (pv_player_finish(mPlayer) != PV_SUCCESS) {
+			meddbg("Error! pv_player_finish failed!\n");
+		}
+		delete mPlayer;
+		mPlayer = nullptr;
 	}
 #endif
 }
 
 size_t Decoder::pushData(unsigned char* buf, size_t size)
 {
 #ifdef CONFIG_AUDIO_CODEC
-	size_t rmax = pv_player_dataspace(&mPlayer);
+	size_t rmax = pv_player_dataspace(mPlayer);
 	if (size > rmax) {
 		meddbg("Error!! data is larger than rmax\n");
 		size = rmax;
 	}
 
-	return pv_player_pushdata(&mPlayer, buf, size);
+	return pv_player_pushdata(mPlayer, buf, size);
 #endif
 	return 0;
 }
 
 bool Decoder::getFrame(unsigned char* buf, size_t* size, unsigned int* sampleRate, unsigned short* channels)
 {
 #ifdef CONFIG_AUDIO_CODEC
-	if (mPlayer.audio_type == 0) {
-		mPlayer.audio_type = _get_audio_type(mPlayer.rbsp);
-		medvdbg("audio_type %d\n", mPlayer.audio_type);
+	if (mPlayer->audio_type == 0) {
```

Comment:
0 means type_unknown, right?

Reply:
Yes. It means that **Run getting audio format when only If the decoder doesn't know**

Reply:
ok then let's use enum value instead of constant.
It is only one thing in this PR

Reply:
What advantage do we get by this movement from stack to heap? Keeping the data in stack is advantageous from multiple instance point of view.  If the data structure was managed in stak, it is easy to have multiple instane of the player. But with heap is it possible to manage multiple instance of player?

Reply:
Now I'm using valgrind to check memory corruption and leak on Linux.
Using stack occurs memory corruption on linux now. So I changed to the heap.
The issue has been fixed but I have to find the detail reason.